# Loan repayment predictor

Data set description

## Project goal
Using the dataset described above, we will make predictions on the data to answer following question: __Will the borrower pay the loan?__

In order to achieve our goal, we will follow following workflow:
- Data exploration and understanding of each column
- Data cleaning
- Feature processing and selection
- Development of prediction models
- Error comparison and model selection

In [1]:
# Disable warnings in Anaconda
import warnings
warnings.filterwarnings('ignore')

import pandas as pd # Data processing
pd.options.display.max_columns = 52

import numpy as np # Linear algebra

import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set_palette("colorblind")

# Helper functions
def print_list(_list,title):
    print(title,'\n')
    for item in _list:
        print('  - ', item)
        
def print_series(_series,title):
    print(title,'\n')
    print(_series.to_string())

In [2]:
loans_2007 = pd.read_csv('loans_2007.csv')

## Data exploration

Before starting with the data cleaning process, we will identify columns with the following attributes:

- Columns that leak information from the future -to be avoided in our model
- Columns that don't impact the borrower's ability to pay off the loan
- Columns that need data cleaning
- Missing data
- Columns that can be combined
- Redundant columns

Our target column is `loan_status`. This column tells us wether the borrower completed the loan payment and under which circumstances.

For more understanding of our data, let's print the first 5 rows:

In [3]:
loans_2007.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501  1296599.0     5000.0       5000.0           4975.0   36 months   
1  1077430  1314167.0     2500.0       2500.0           2500.0   60 months   
2  1077175  1313524.0     2400.0       2400.0           2400.0   36 months   
3  1076863  1277178.0    10000.0      10000.0          10000.0   36 months   
4  1075358  1311748.0     3000.0       3000.0           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-2011   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-2011  Charged Off   
2           RENT     12252.0        Not Verified  Dec-2011   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-2011   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-2011      Current   

  pymnt_plan         purpose                 title zip_code addr_state    dti  \
0          n     credit_card              Computer    860xx         AZ  27.65   
1          n             car                  bike    309xx         GA   1.00   
2          n  small_business  real estate business    606xx         IL   8.72   
3          n           other              personel    917xx         CA  20.00   
4          n           other              Personal    972xx         OR  17.94   

   delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal  \
0          0.0         Jan-1985             1.0       3.0      0.0    13648.0   
1          0.0         Apr-1999             5.0       3.0      0.0     1687.0   
2          0.0         Nov-2001             2.0       2.0      0.0     2956.0   
3          0.0         Feb-1996             1.0      10.0      0.0     5598.0   
4          0.0         Jan-1996             0.0      15.0      0.0    27783.0   

  revol_util  total_acc initial_list_status  out_prncp  out_prncp_inv  \
0      83.7%        9.0                   f       0.00           0.00   
1       9.4%        4.0                   f       0.00           0.00   
2      98.5%       10.0                   f       0.00           0.00   
3        21%       37.0                   f       0.00           0.00   
4      53.9%       38.0                   f     461.73         461.73   

    total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
0   5863.155187          5833.84          5000.00         863.16   
1   1008.710000          1008.71           456.46         435.17   
2   3005.666844          3005.67          2400.00         605.67   
3  12231.890000         12231.89         10000.00        2214.92   
4   3581.120000          3581.12          2538.27        1042.85   

   total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
0                0.00        0.00                     0.00     Jan-2015   
1                0.00      117.08                     1.11     Apr-2013   
2                0.00        0.00                     0.00     Jun-2014   
3               16.97        0.00                     0.00     Jan-2015   
4                0.00        0.00                     0.00     Jun-2016   

   last_pymnt_amnt last_credit_pull_d  collections_12_mths_ex_med  \
0           171.62           Jun-2016                         0.0   
1           119.66           Sep-2013                         0.0   
2           649.91           Jun-2016                         0.0   
3           357.48           Apr-2016                     

In [4]:
print('Number of rows: ',loans_2007.shape[0])
print('Number of columns: ',loans_2007.shape[1])

Number of rows:  42538
Number of columns:  52


## Data cleaning

### Eliminating irrelevant columns and treating missing values

In [5]:
relevant_columns = ['loan_amnt','term','int_rate','installment',
                   'emp_length','home_ownership','annual_inc',
                   'pymnt_plan','purpose','addr_state','delinq_2yrs',
                   'earliest_cr_line','inq_last_6mths',
                    'open_acc','pub_rec','total_acc','application_type',
                    'tax_liens','verification_status','title','dti',
                    'revol_bal','revol_util','last_credit_pull_d', 
                    'pub_rec_bankruptcies',                 
                    'loan_status' #target col
                   ]
# grade and sub_grade not included since they include redundant info from int_rate
# earliest_cr_line : create new feature: years since first credit line
#inq_last_ [6 and 12] months: figure sth out
relevant_columns
drop_columns = loans_2007.columns.difference(relevant_columns)
print_list(drop_columns.tolist(),'Columns to drop: 1st round')

Columns to drop: 1st round 

  -  acc_now_delinq
  -  chargeoff_within_12_mths
  -  collection_recovery_fee
  -  collections_12_mths_ex_med
  -  delinq_amnt
  -  emp_title
  -  funded_amnt
  -  funded_amnt_inv
  -  grade
  -  id
  -  initial_list_status
  -  issue_d
  -  last_pymnt_amnt
  -  last_pymnt_d
  -  member_id
  -  out_prncp
  -  out_prncp_inv
  -  policy_code
  -  recoveries
  -  sub_grade
  -  total_pymnt
  -  total_pymnt_inv
  -  total_rec_int
  -  total_rec_late_fee
  -  total_rec_prncp
  -  zip_code


In [6]:
loans_2007=loans_2007.drop(drop_columns,axis=1) #drop columns

In [7]:
print_series(loans_2007['loan_status'].value_counts(),'Types of loan status - before filtering data')

Types of loan status - before filtering data 

Fully Paid                                             33136
Charged Off                                             5634
Does not meet the credit policy. Status:Fully Paid      1988
Current                                                  961
Does not meet the credit policy. Status:Charged Off      761
Late (31-120 days)                                        24
In Grace Period                                           20
Late (16-30 days)                                          8
Default                                                    3


In [8]:
# keep loans fully paid or charged off only
loans_2007 = loans_2007[
    (loans_2007['loan_status'] == 'Fully Paid') |
    (loans_2007['loan_status'] == 'Charged Off') 
     ]

# transform fully paid and charged off to numeric values
mapping_dict = {
    'loan_status': {
        'Fully Paid': 1,
        'Charged Off': 0
    }
}
loans_2007 = loans_2007.replace(mapping_dict)

print_series(loans_2007['loan_status'].value_counts(),'Types of loan status - fully paid = 1, charged off = 0')

Types of loan status - fully paid = 1, charged off = 0 

1    33136
0     5634


There are way more paid off than charged off loans, which means that our model will tend to predict more 1 (paid off) loans than charged off - biased model.

In [9]:
loans_2007['tax_liens'].unique() # drop this column, not relevant for predictions

array([ 0., nan])

In [10]:
# drop columns that contain only 1 unique value besides null

drop_columns = []

for col in loans_2007.columns:
    non_null_vals = loans_2007[col].dropna()
    unique_vals = non_null_vals.unique()
    num_unique_vals = len(unique_vals)
    if num_unique_vals == 1:
        drop_columns.append(col)

loans_2007 = loans_2007.drop(drop_columns,axis=1)

print_list(drop_columns,'Columns to drop: 2nd round: We will drop columns that have one unique value')

Columns to drop: 2nd round: We will drop columns that have one unique value 

  -  pymnt_plan
  -  application_type
  -  tax_liens


In [11]:
null_vals = loans_2007.isnull().sum()

print_series(null_vals[null_vals > 0],'List of null values')

List of null values 

emp_length              1036
title                     11
revol_util                50
last_credit_pull_d         2
pub_rec_bankruptcies     697


In [12]:
print_series(loans_2007['pub_rec_bankruptcies'].value_counts(),'Value count: pub_rec_bankruptcies')

Value count: pub_rec_bankruptcies 

0.0    36422
1.0     1646
2.0        5


As we can see above, `pub_rec_bankruptcies` offers little to no variability. We'll drop this column.

In [13]:
loans_2007.drop(['pub_rec_bankruptcies'],axis=1,inplace=True)

In [14]:
print('Missing emp_length values: {:.2f}% of total values.'.format(
      100 * null_vals['emp_length']/len(loans_2007)))

Missing emp_length values: 2.67% of total values.


There are 1036 missing values in the column `emp_length`. Although we could impute these values using the mode, we can afford dropping all the rows with na values, since they represent less than a 3% of the total number of rows.

In [15]:
loans_2007.dropna(inplace=True)
loans_2007 = loans_2007.reset_index(drop=True) # reset indexes
print_series(loans_2007.isnull().sum(),"Missing values count - cleaned data")

Missing values count - cleaned data 

loan_amnt              0
term                   0
int_rate               0
installment            0
emp_length             0
home_ownership         0
annual_inc             0
verification_status    0
loan_status            0
purpose                0
title                  0
addr_state             0
dti                    0
delinq_2yrs            0
earliest_cr_line       0
inq_last_6mths         0
open_acc               0
pub_rec                0
revol_bal              0
revol_util             0
total_acc              0
last_credit_pull_d     0


## Feature processing and selection
Our machine learning model will only work with numeric columns. Since we have valuable columns with categorical values, we will proceed by creating dummy variables for these. We will also eliminate categorical features that contain an excesive number of unique values, since this would require too many dummy variables and would slow our model down.

In [16]:
print_series(loans_2007.dtypes.value_counts(),"Column types")

Column types 

object     11
float64    10
int64       1


In [17]:
object_cols_df = loans_2007.select_dtypes(include='object')
object_cols_df.head(1)

term int_rate emp_length home_ownership verification_status  \
0   36 months   10.65%  10+ years           RENT            Verified   

       purpose     title addr_state earliest_cr_line revol_util  \
0  credit_card  Computer         AZ         Jan-1985      83.7%   

  last_credit_pull_d  
0           Jun-2016

As seen above, we have 11 columns that contain text values. We will proceed as follows:

__Drop columns:__

- `last_credit_pull_d`: Requires exhaustive feature engineering
- `addr_state`: Has too many values
- `title`': Has too many values
- `earliest_cr_line`: Requires exhaustive feature engineering

The rest of the columns will be treated, either switching the string values to numeric values (one number per value) or eliminating any unnecessary string values -i.e. percent symbol.

Our cleaned data, before creating dummy variables, looks as follows:

In [18]:
drop_cols = ['last_credit_pull_d', #requires exhaustive feature engineering
             'addr_state', #would require 49 dummy variables
             'title', #too many values
             'earliest_cr_line' #requires exhaustive feature engineering
            ]
loans_2007.drop(drop_cols,axis=1,inplace=True)

loans_2007['int_rate']=loans_2007['int_rate'].str.rstrip('%')
loans_2007['int_rate']=loans_2007['int_rate'].astype('float')

loans_2007['revol_util']=loans_2007['revol_util'].str.rstrip('%')
loans_2007['revol_util']=loans_2007['revol_util'].astype('float')

mapping_emp_length = { #converting to numeric values instead of categories
    "emp_length": {
        "10+ years": 10,
        "9 years": 9,
        "8 years": 8,
        "7 years": 7,
        "6 years": 6,
        "5 years": 5,
        "4 years": 4,
        "3 years": 3,
        "2 years": 2,
        "1 year": 1,
        "< 1 year": 0,
        "n/a": 0
    }
}

loans_2007=loans_2007.replace(mapping_emp_length)

loans_2007.head()

loan_amnt        term  int_rate  installment emp_length home_ownership  \
0     5000.0   36 months     10.65       162.87         10           RENT   
1     2500.0   60 months     15.27        59.83          0           RENT   
2     2400.0   36 months     15.96        84.33         10           RENT   
3    10000.0   36 months     13.49       339.31         10           RENT   
4     5000.0   36 months      7.90       156.46          3           RENT   

   annual_inc verification_status  loan_status         purpose    dti  \
0     24000.0            Verified            1     credit_card  27.65   
1     30000.0     Source Verified            0             car   1.00   
2     12252.0        Not Verified            1  small_business   8.72   
3     49200.0     Source Verified            1           other  20.00   
4     36000.0     Source Verified            1         wedding  11.20   

   delinq_2yrs  inq_last_6mths  open_acc  pub_rec  revol_bal  revol_util  \
0          0.0             1.0       3.0      0.0    13648.0        83.7   
1          0.0             5.0       3.0      0.0     1687.0         9.4   
2          0.0             2.0       2.0      0.0     2956.0        98.5   
3          0.0             1.0      10.0      0.0     5598.0        21.0   
4          0.0             3.0       9.0      0.0     7963.0        28.3   

   total_acc  
0        9.0  
1        4.0  
2       10.0  
3       37.0  
4       12.0

### Creating dummy variables
We will create dummy variables for our categorical columns:
- `home_ownership`
- `verification_status`
- `purpose`
- `term`

Once we create them, we will drop the text columns.

In [19]:
dummy_df = pd.get_dummies(loans_2007[[
    'home_ownership', 'verification_status', 'purpose', 'term'
]])
loans_2007 = pd.concat([loans_2007,dummy_df],axis=1) #adding dummy cols to df

drop_cols = ['home_ownership', 'verification_status', 'purpose', 'term']
loans_2007.drop(drop_cols,axis=1, inplace=True) #drop text cols

## Development of prediction models

### Error metric

In [20]:
possibilities = {
    "loan_status": [0,0,1,1],
    "prediction": [1,0,1,0],
    "error": ["False Positive",
           "True Negative",
           "True Positive",
           "False Negative"]
}
print("Possible outcomes of a binary classification prediction:\n")
print(pd.DataFrame(possibilities).to_string(index=False))

Possible outcomes of a binary classification prediction:

 loan_status  prediction           error
           0           1  False Positive
           0           0   True Negative
           1           1   True Positive
           1           0  False Negative


__Loan status - possible outcomes__:
- 1: Loan will be paid off on time.
- 0: Loan will not be paid off on time.

__Prediction errors__:
- *False Positive*: Our model predicts 1 (will be paid off on time) but the actual status is 0 (not paid off on time). __This is our worst case scenario, since this approving such a loan would make us lose money.__

- *False Negative*: Our model predicts 0 (not paid off on time) but the actual status is 1 (will be paid off on time). 

__Class imbalance__:

Since we have way more loans that were paid off on time (1 values in our target `loans_2007['loan_status']`), our model will be more prone to predict that a loan will paid off on time. For this reason, using accuracy to measure the error metric is not the best approach.

To determine our error metric, we will look at following parameters:

__(rewrite formulas and explain)__
- *High recall/True positive rate*: $ tpr = tp / (tp + fn) $
- *Low fall-out/False positive rate*: $False_prate = false_p / (false_p + true_n)$

The higher our recall, the better our model will predict positive cases -we will grant more loans that will be paid off, and the lower our fall-out, the better our model will predict negative cases -we will reject more loans that would not be paid off. These two ratios are correlated and will fluctuate. 

Optimizing these ratios will yield the highest chances of gaining (or not losing) money when using our model to approve or reject a loan.

## Logistic regression

K-Fold cross validation.


In order to work around our class imbalance, we will penalize some rows by setting the `class_weight` parameter to `balanced` in our LogisticRegression instance. Our penalty will be inversely proportional to the class frequencies.

![Diagram of k-fold cross validation](https://upload.wikimedia.org/wikipedia/commons/thumb/b/b5/K-fold_cross_validation_EN.svg/521px-K-fold_cross_validation_EN.svg.png)

_Diagram of k-fold cross validation_

In [26]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import accuracy_score

features = loans_2007.drop(['loan_status'],axis=1) # drop target column
target = loans_2007["loan_status"]

# Create a LogisticRegression instance 
# balanced class_weight will penalize loan_status = 1 -higher frequency
lr = LogisticRegression(class_weight='balanced')

k = 3

predictions = cross_val_predict(lr, features, target, cv=k)
predictions = pd.Series(predictions)

# True negatives
tn_bool_mask = loans_2007[(predictions == 0) & (loans_2007["loan_status"] == 0)]
tn = tn_bool_mask.shape[0]
# True positives
tp_bool_mask = loans_2007[(predictions == 1) & (loans_2007["loan_status"] == 1)]
tp = tp_bool_mask.shape[0]
# False negatives
fn_bool_mask = loans_2007[(predictions == 0) & (loans_2007["loan_status"] == 1)]
fn = fn_bool_mask.shape[0]
# False positives
fp_bool_mask = loans_2007[(predictions == 1) & (loans_2007["loan_status"] == 0)]
fp = fp_bool_mask.shape[0]

# Computing the true positive rate
tpr = tp/(tp+fn)

# Computing the false positive rate
fpr = fp/(fp + tn)

# Accuracy score - used only for model comparison purposes
accuracy = accuracy_score(loans_2007["loan_status"], predictions)
print("Logistic regression model, k = {}, class_weight = {}".format(k,'balanced'), '\n')
print("True positive rate: {:.2f}%".format(tpr*100))
print("False positive rate: {:.2f}%".format(fpr*100)) 
print("Number of false positives: {}".format(fp))
print("Accuracy score: {:.2f}%".format(accuracy*100))

Logistic regression model, k = 3, class_weight = balanced 

True positive rate: 67.02%
False positive rate: 38.86%
Number of false positives: 2094
Accuracy score: 66.18%


## Random forest classifier model

In [27]:
from sklearn.ensemble import RandomForestClassifier

rfc = RandomForestClassifier(random_state=1,class_weight='balanced')

predictions = cross_val_predict(rfc, features, target, cv=3)
predictions = pd.Series(predictions)

# True negatives
tn_bool_mask = loans_2007[(predictions == 0) & (loans_2007["loan_status"] == 0)]
tn = tn_bool_mask.shape[0]
# True positives
tp_bool_mask = loans_2007[(predictions == 1) & (loans_2007["loan_status"] == 1)]
tp = tp_bool_mask.shape[0]
# False negatives
fn_bool_mask = loans_2007[(predictions == 0) & (loans_2007["loan_status"] == 1)]
fn = fn_bool_mask.shape[0]
# False positives
fp_bool_mask = loans_2007[(predictions == 1) & (loans_2007["loan_status"] == 0)]
fp = fp_bool_mask.shape[0]

# Computing the true positive rate
tpr = tp/(tp+fn)

# Computing the false positive rate
fpr = fp/(fp + tn)

# Accuracy score - used only for model comparison purposes
accuracy = accuracy_score(loans_2007["loan_status"], predictions)

print("Random forest classifier model, k = {}, class_weight = {}".format(k,'balanced'), '\n')
print("True positive rate: {:.2f}%".format(tpr*100))
print("False positive rate: {:.2f}%".format(fpr*100)) 
print("Number of false positives: {}".format(fp))
print("Accuracy score: {:.2f}%".format(accuracy*100))

Random forest classifier model, k = 3, class_weight = balanced 

True positive rate: 97.01%
False positive rate: 91.82%
Number of false positives: 4948
Accuracy score: 84.30%


## Error comparison and model selection
